Doc/library/typing.rst

```diff
@@ -2429,6 +2429,72 @@ Functions and decorators
 
    .. versionadded:: 3.11
 
+.. decorator:: dataclass_transform
+
+   The :data:`~typing.dataclass_transform` annotation may be used to
+   decorate a function that is itself a decorator, a class, or a metaclass.
+   The presence of ``@dataclass_transform()`` tells a static type checker that the
+   decorated function, class, or metaclass performs runtime "magic" that
+   transforms a class, endowing it with dataclass-like behaviors.
+
+   Example usage with a decorator function::
+
+      _T = TypeVar("_T")
+
+      @dataclass_transform()
+      def create_model(cls: type[_T]) -> type[_T]:
+          ...
+          return cls
+
+      @create_model
+      class CustomerModel:
+          id: int
+          name: str
+
+   On a base class::
+
+      @dataclass_transform()
+      class ModelBase: ...
+
+      class CustomerModel(ModelBase):
+          id: int
+          name: str
+
+   On a metaclass::
+
+      @dataclass_transform()
+      class ModelMeta(type): ...
+
+      class ModelBase(metaclass=ModelMeta): ...
+
+      class CustomerModel(ModelBase):
+          id: int
+          name: str
+
+   Each of the ``CustomerModel`` classes defined in this example will now
+   behave similarly to a dataclass created with the ``@dataclasses.dataclass``
+   decorator. For example, the type checker will synthesize an ``__init__``
+   method.
```

Comment:
May as well cross-link to the dataclass docs here. Also, make it clear that this has no behaviour at runtime.

```suggestion
   be treated by a type-checker similarly to a dataclass created with the :func:`@dataclasses.dataclass <dataclasses.dataclass>`
   decorator. For example, the type checker will assume that an ``__init__``
   method will be autogenerated by the runtime.
```